python/tests/ibd.py

```diff
@@ -0,0 +1,485 @@
+            while S_head is not None:
+                u = S_head.node
+                list_to_add = SegmentList()
+                if u in self.samples:
+                    list_to_add + Segment(S_head.left, S_head.right, u)
+                else:
+                    s = self.A[u].head
+                    while s is not None:
+                        intvl = (max(S_head.left, s.left), min(S_head.right, s.right))
+                        if intvl[1] - intvl[0] > 0:
+                            list_to_add + Segment(intvl[0], intvl[1], s.node)
+                        s = s.next
+                self.add_to_chain(list_to_add)
+                S_head = S_head.next
+
+            # If we wanted to squash segments, we'd do it here.
+
+            # Use the info in SegL to find new ibd segments that descend from the
+            # parent node currently being processed.
+            if self.SegL_head.next is not None or self.current_parent in self.samples:
+                nodes_to_remove = self.calculate_ibd_segs()
+
+                # g. Remove elements of the list of ancestral segments A if they
+                # are no longer needed. (Memory-pruning step)
+                for n in nodes_to_remove:
+                    if self.current_parent in self.parent_list[n]:
+                        self.parent_list[n].remove(self.current_parent)
+                    if len(self.parent_list[n]) == 0:
+                        self.A[n] = []
+
+            # Create the list of ancestral segments, A, descending from current node.
+            # Concatenate all of the segment lists in SegL into a single list, and
+            # store in A. This is a list of all sample segments descending from
+            # the currently processed node.
+            seglist = self.SegL_head
+            while seglist is not None:
+                self.A[self.current_parent] + seglist
+                seglist = seglist.next
+
+            # Clear the temporary segment lists.
+            self.SegL_head = SegmentList()
+            self.SegL_tail = SegmentList()
+
+        return self.ibd_segments
+
+    def calculate_ibd_segs(self):
+        """
+        Calculates all new IBD segments found at the current iteration of the
+        algorithm. Returns information about nodes processed in this step, which
+        allows memory to be cleared as the procedure runs.
+        """
+
+        list0 = self.SegL_head
+        # Iterate through all pairs of segment lists.
+        while list0.next is not None:
+            list1 = list0.next
+            while list1 is not None:
+                seg0 = list0.head
+                # Iterate through all segments with one taken from each list.
+                while seg0 is not None:
+                    seg1 = list1.head
+                    while seg1 is not None:
+                        if seg0.node == seg1.node:
+                            seg1 = seg1.next
+                            continue
```

Comment:
this isn't needed, as any two segments with the same node shouldn't overlap, so will `continue` five lines down anyhow?

Reply:
yup, good spot!